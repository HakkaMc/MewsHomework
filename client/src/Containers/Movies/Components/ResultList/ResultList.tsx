import React from "react";
import { connect } from "react-redux";
import { List, ListItem, ListItemText } from "@material-ui/core";
import { compose } from "recompose";
import get from "lodash/get";

import { Movie } from "../../../../reducers/Movies/model";
import { ReduxState, actions } from "../../../../reducers";

export interface Props {
	clearSelectedMovie: typeof actions.movies.clearSelectedMovie;
	movies: Array<Movie>;
	selectedMovie?: Movie;
	setSelectedMovie: typeof actions.movies.setSelectedMovie;
	getMovieDetail: typeof actions.movies.getMovieDetail;
	searchMovies: typeof actions.movies.searchMovies;
	page: number;
	totalResults: number;
	totalPages: number;
}

export const ResultList: React.FC<Props> = (props: Props) => {
	const getMovieDetail = (movieId: number) => () => {
		props.clearSelectedMovie();
		props.getMovieDetail(movieId);
	};

	const list: Array<any> = (props.movies || []).map((movie: any) => {
		return (
			<ListItem
				button
				key={movie.id}
				selected={movie.id === get(props, "selectedMovie.id")}
				onClick={getMovieDetail(movie.id)}
			>
				<ListItemText primary={movie.title} />
			</ListItem>
		);
	});

	return (
		<List component="nav" style={{ paddingTop: 0 }}>
			{list}
		</List>
	);
};

export default compose<Props, any>(
	connect(
		(state: ReduxState) => ({
			page: state.movies.page,
			totalResults: state.movies.totalResults,
			totalPages: state.movies.totalPages,
			movies: state.movies.movies,
			selectedMovie: state.movies.selectedMovie
		}),
		{
			setSelectedMovie: actions.movies.setSelectedMovie,
			getMovieDetail: actions.movies.getMovieDetail,
			searchMovies: actions.movies.searchMovies,
			clearSelectedMovie: actions.movies.clearSelectedMovie
		}
	)
)(ResultList);
