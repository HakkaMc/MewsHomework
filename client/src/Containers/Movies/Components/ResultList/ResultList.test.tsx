import React from "react";
import { shallow } from "enzyme";

import { ResultList, Props } from "./ResultList";
import { ListItem } from "@material-ui/core";

describe("<ResultList/>", () => {
	let props: Props;

	beforeEach(() => {
		props = {
			clearSelectedMovie: jest.fn(),
			movies: [
				{
					id: 0,
					original_language: "en",
					original_title: "title",
					poster_path: "",
					title: "title",
					vote_average: 0,
					vote_count: 0
				}
			],
			selectedMovie: undefined,
			setSelectedMovie: jest.fn(),
			getMovieDetail: jest.fn(),
			searchMovies: jest.fn(),
			page: 1,
			totalResults: 2,
			totalPages: 1
		};
	});

	it("Match snapshot with valid mock data", () => {
		props.totalPages = 0;
		const wrapper = shallow(<ResultList {...props} />);
		expect(wrapper.isEmptyRender()).toBe(false);
		expect(wrapper).toMatchSnapshot();
	});

	it("Selecting an  item calls clearSelectedMovie and getMovieDetail", () => {
		props.totalPages = 0;
		const wrapper = shallow(<ResultList {...props} />);
		// @ts-ignore
		wrapper.find(ListItem).prop("onClick")(props.movies[0].id);
		expect(props.clearSelectedMovie).toHaveBeenCalled();
		expect(props.getMovieDetail).toHaveBeenCalledWith(props.movies[0].id);
	});
});
