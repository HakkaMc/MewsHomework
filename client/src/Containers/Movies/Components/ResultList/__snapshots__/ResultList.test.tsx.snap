// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultList/> Match snapshot with valid mock data 1`] = `
<WithStyles(ForwardRef(List))
  component="nav"
  style={
    Object {
      "paddingTop": 0,
    }
  }
>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    key="0"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemText))
      primary="title"
    />
  </WithStyles(ForwardRef(ListItem))>
</WithStyles(ForwardRef(List))>
`;
